### PR TITLE
This PR include ecf directory.

### PR DESCRIPTION
**Pull request by lgannoaa**
As directed by Rahul Mahajan this ecflow PR will include "PBS -l place" from the Google sheet column L (snapshot at 5:00p January 5, 2022) and "PBS -l walltime" from current operation.  This PR will NOT be tested at time of submit.

The snapshot URL https://docs.google.com/spreadsheets/d/1sef7gAq5OZAP_ioySKKwcMmmOkhlM0RMO67XTIoMlDg/edit?usp=sharing

gldas is updated to gldas.fd (e757e6ed)
  PBS -l for job gdas gldas has been modified to:
    #PBS -l place=vscatter,select=1:mpiprocs=128:ompthreads=1:ncpus=128

This PR is only for update under the $HOMEgfs/ecf location for wcoss2 GFS transition.

### CODE REVIEW

File: ecf/include/head.h [hunk 1]
```diff
@@ -4,45 +4,83 @@ export PS4='+ $SECONDS + '
...
-    module load EnvVars/$EnvVars_ver
-    module load ecflow/$ecflow_ver
-    POST_OUT=/var/lsf/ecflow_post_in.$USER.$LSB_BATCH_JID
+export ECF_RID=${ECF_RID:-${PBS_JOBID:-$$}}
+export ECF_JOB=%ECF_JOB%
+export ECF_JOBOUT=%ECF_JOBOUT%
+export ecflow_ver=%ecflow_ver%
+
+if [ -d /apps/ops/prod ]; then # On WCOSS2
+####  set +x
```
Comment (aerorahul) [inline ecf/include/head.h:17]: please retain this `set +x` and `set -x` on line 27.

File: ecf/include/head.h [hunk 2]
```diff
@@ -4,45 +4,83 @@ export PS4='+ $SECONDS + '
...
+  echo "Running 'module reset'"
+  module reset
+  module load envvar/1.0
+  module load PrgEnv-intel/8.1.0
+  module load craype/2.7.8
+  module load intel/19.1.3.304
+  module load cray-mpich/8.1.9
+  module load cray-pals/1.0.17
+  module load python/3.8.6
```
Comment (aerorahul) [inline ecf/include/head.h:26]: Curious, why `python` is required in `head.h`?

File: ecf/include/head.h [hunk 3]
```diff
@@ -4,45 +4,83 @@ export PS4='+ $SECONDS + '
...
 
-cat > $POST_OUT <<ENDFILE
-ECF_NAME=${ECF_NAME}
-ECF_HOST=${ECF_HOST}
-ECF_PORT=${ECF_PORT}
-ECF_PASS=${ECF_PASS}
-ECF_TRYNO=${ECF_TRYNO}
-ECF_RID=${ECF_RID}
-ENDFILE
+export HOMEgfs=/lfs/h2/emc/global/noscrub/Lin.Gan/git/feature-ops-wcoss2
```
Comment (aerorahul) [inline ecf/include/head.h:30]: Shouldn't this have a `Developer Overwrite` wrap around it?

File: ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf [hunk 4]
```diff
@@ -29,14 +21,14 @@ export model=gfs
...
 # Load modules
 ############################################################
-module load lsf/${lsf_ver}
-module load impi/${impi_ver}
-module load CFP/${CFP_ver}
-module load python/${python_ver}
-module load NetCDF-parallel/${netcdf_parallel_ver}
-module load HDF5-parallel/${hdf5_parallel_ver}
+module load netcdf/${netcdf_ver}
+module load hdf5/${hdf5_ver}
```
Comment (aerorahul) [inline ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:25]: This order needs to be reversed.

File: ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf [hunk 5]
```diff
@@ -29,14 +21,14 @@ export model=gfs
...
+module load cfp/${cfp_ver}
 module load crtm/${crtm_ver}
+module load libjpeg/${libjpeg_ver}
 module load grib_util/${grib_util_ver}
+module load wgrib2/${wgrib2_ver}
```
Comment (aerorahul) [inline ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf]: Where is `wgrib2` used in the analysis job?

File: ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf (hunk 4 above)
Comment (aerorahul) [reply ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:25]: This comment is for everywhere the netcdf and hdf5 modules are loaded.
netcdf depends on hdf5 and the latter needs to be loaded before.

It is likely there are other module loads that are not following their dependency tree.
Reviewing that will take a long time at this point.

File: ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf [hunk 6]
```diff
@@ -29,14 +21,14 @@ export model=gfs
 ############################################################
 # Load modules
 ############################################################
-module load lsf/${lsf_ver}
-module load impi/${impi_ver}
-module load CFP/${CFP_ver}
-module load python/${python_ver}
-module load NetCDF-parallel/${netcdf_parallel_ver}
-module load HDF5-parallel/${hdf5_parallel_ver}
+module load netcdf/${netcdf_ver}
```
Comment (aerorahul) [inline ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:24]: Where are the PrgIntel, intel, mpich, cray-pals etc modules loaded?
I thought they are needed to execute.

I see in jgdas_atmos_dump.ecf, some of these are loaded.
So I wonder why the inconsistency?

**Review by aerorahul — COMMENTED** [review]
It is going to be difficult to review this in 1 sitting.
So here are my first comments looking at the first few files.

File: ecf/include/head.h (hunk 2 above)
Comment (aerorahul) [reply ecf/include/head.h:26]: Also.  Does every job need the mpich module?
By that extension, we could pretty much emulate `load_fv3gfs_modules.sh` in `head.h` and not have to load any module in any of the `ecf` scripts.

File: ecf/scripts/gdas/atmos/obsproc/dump/jgdas_atmos_tropcy_qc_reloc.ecf [hunk 7]
```diff
@@ -1,16 +1,13 @@
...
-#BSUB -P %PROJ%
-#BSUB -J %E%gdas_atmos_tropcy_qc_reloc_%CYC%
-#BSUB -o %COM%/output/%ENVIR%/today/gdas_atmos_tropcy_qc_reloc_%CYC%.o%J
-#BSUB -W 0:05
-#BSUB -n 1
-#BSUB -R affinity[core(1)]
-#BSUB -R span[ptile=1]
-#BSUB -cwd /tmp
+#PBS -S /bin/bash
+#PBS -N %E%gdas_atmos_tropcy_qc_reloc_%CYC%
```
Comment (KateFriedman-NOAA) [inline ecf/scripts/gdas/atmos/obsproc/dump/jgdas_atmos_tropcy_qc_reloc.ecf]: We are supposed to remove the `%E%` text from the jobname line:
```
#PBS -N gdas_atmos_tropcy_qc_reloc_%CYC%
```

See the request to remove `%E%` in the NCO checklist for hand-off:
https://docs.google.com/document/d/13p_wvsKTejk7M5q2KMyOeijH4shghlJrTpQlZkThtDI/edit#

Please remove `%E%` from all ecf scripts, thanks!

**Comment by KateFriedman-NOAA** [comment]
As mentioned in another comment above...we are supposed to remove the `%E%` text from the jobname line. We are also supposed to make sure that "job name (-N) start with package name in PBSpro submission". See both requirements from NCO in this checklist for hand-off:
https://docs.google.com/document/d/13p_wvsKTejk7M5q2KMyOeijH4shghlJrTpQlZkThtDI/edit#

Please address the needed changes and other questions I tagged you on in that checklist document, thanks!

**Comment by WalterKolczynski-NOAA** [comment]
Please remember that the PR descriptions become the commit message. Statements like "As directed by..." do not really belong in the git log (and definitely not in the commit title). If necessary they can be added as a comment on the PR. Please use good commit message practices when writing commit messages (see the wiki).

**Comment by lgannoaa** [comment]
Here is a reply for all question above:
1. %E% has been removed.
2. analysis jobs do not use wgrib2. It removed and under testing.
3. Module position is not been changed in over two weeks. The output has been reviewed by code managers during the 20210824 testing run. Prefer not to change at this time without testing.
4. NCO does not use the ecf/include directory. They have their production version that handle all models.
5. NCO require all job script in ecflow load its own model. It is not allowed to load all modules on every jobs.

File: ecf/defs/prod12.def [hunk 8]
```diff
@@ -3,30 +3,38 @@ extern /prod06/gdas/atmos/post
 extern /prod06/gdas/enkf/post
 #### ecen
 extern /prod06/gdas/atmos/post_processing/jgdas_atmos_chgres_forenkf
-
+extern /prod18/gdas/atmos/analysis/jgdas_atmos_analysis
+extern /prod18/gfs/atmos/analysis/jgfs_atmos_analysis
+extern /prod18/gdas/enkf/analysis/recenter/ecen
+extern /totality_limit:TOTALITY
```
Comment (aerorahul) [inline ecf/defs/prod12.def:9]: Is this not necessary in the other `prod00.def`, `prod06.def` and `prod18.def` files?

File: ecf/defs/prod00.def [hunk 9]
```diff
@@ -3,30 +3,38 @@ extern /prod18/gdas/atmos/post
 extern /prod18/gdas/enkf/post
 #### ecen
 extern /prod18/gdas/atmos/post_processing/jgdas_atmos_chgres_forenkf
+extern /prod06/gdas/atmos/analysis/jgdas_atmos_analysis
+extern /prod06/gfs/atmos/analysis/jgfs_atmos_analysis
+extern /prod06/gdas/enkf/analysis/recenter/ecen
```
Comment (aerorahul) [inline ecf/defs/prod00.def:8]: I am curious what in the 00z suite is depending on a trigger from 06z.
I don't see any references to triggers from these.

As a matter of fact, the only triggers are from 
`/prod18/gdas/enkf/post` and `/prod18/gdas/atmos/post_processing/jgdas_atmos_chgres_forenkf`

**Comment by WalterKolczynski-NOAA** [comment]
There are a bunch of places where ops settings have been replaced by dev ones. I know they are needed for us to run it, but we are very close to hand-off, so I don't know whether I should be pointing them all out or not.

File: ecf/defs/prod00.def (hunk 9 above)
Comment (WalterKolczynski-NOAA) [reply ecf/defs/prod00.def:8]: I had the same question. Why are new triggers being added when this is only supposed to be a port? And why triggers from three cycles previous?

**Pull request closed without merging**